src/github.com/getlantern/borda/borda.go

```diff
@@ -0,0 +1,246 @@
+	*Options
+	indexedDimensions map[string]bool
+	in                chan *Measurement
+	finalError        eventual.Value
+}
+
+// NewCollector creates and starts a new Collector
+func NewCollector(opts *Options) Collector {
+	if opts.BatchSize == 0 {
+		opts.BatchSize = 1000
+	}
+	if opts.MaxBatchWindow == 0 {
+		opts.MaxBatchWindow = time.Duration(math.MaxInt64)
+	}
+	if opts.MaxRetries == 0 {
+		opts.MaxRetries = 10
+	}
+	if opts.RetryInterval == 0 {
+		opts.RetryInterval = 5 * time.Second
+	}
+
+	c := &collector{
+		Options:           opts,
+		indexedDimensions: make(map[string]bool, len(opts.IndexedDimensions)),
+		in:                make(chan *Measurement, opts.BatchSize*2),
+		finalError:        eventual.NewValue(),
+	}
+
+	glog.Infof("IndexedDimensions: %v", opts.IndexedDimensions)
+	for _, dim := range opts.IndexedDimensions {
+		c.indexedDimensions[dim] = true
+	}
+
+	go c.run()
+	return c
+}
+
+func (c *collector) Submit(m *Measurement) {
+	c.in <- m
+}
+
+func (c *collector) Wait(timeout time.Duration) error {
+	err, _ := c.finalError.Get(timeout)
+	if err != nil {
+		return err.(error)
+	}
+	return nil
+}
+
+// Run runs the Collector. This method returns an error if something goes wrong
+// wrong while starting the Collector.   and an eventual runError on which the caller can block to
+// find out if anything goes wrong while running.
+func (c *collector) run() {
+	timer := time.NewTimer(c.MaxBatchWindow)
+	var batch client.BatchPoints
+	batchSize := 0
+
+	newBatch := func() error {
+		var err error
+		batch, err = client.NewBatchPoints(client.BatchPointsConfig{
+			Database: c.DBName,
+		})
+		if err != nil {
+			return fmt.Errorf("Unable to create batch: %v", err)
+		}
+		batchSize = 0
+		return nil
+	}
+
+	commitBatch := func() error {
+		defer timer.Reset(c.MaxBatchWindow)
+		if batchSize == 0 {
+			return nil
+		}
+		retries := 0
+		for {
+			err := c.WriteToDatabase(batch)
+			if err == nil {
+				return newBatch()
+			}
+			if retries >= c.MaxRetries {
+				return fmt.Errorf("Unable to commit batch, not retrying: %v", err)
+			}
+			glog.Errorf("Unable to commit batch, retrying: %v", err)
+			retries++
+			time.Sleep(c.RetryInterval)
+		}
+	}
+
+	err := newBatch()
+	if err != nil {
+		c.terminate(fmt.Errorf("Unable to create batch: %v", err))
+		return
+	}
+
+	for {
+		select {
+		case m := <-c.in:
+			// Create a point for the original measurement key and a point for the
+			// combined measurement.
+			tags := make(map[string]string, len(c.IndexedDimensions))
+			fields := make(map[string]interface{}, len(m.Values)+len(m.Dimensions))
+			addTagOrField := func(key string, value interface{}, isDimension bool) {
+				if value != nil && value != "" {
+					if isDimension && c.indexedDimensions[key] {
+						var stringValue string
+						switch v := value.(type) {
+						case string:
+							stringValue = v
+						default:
+							stringValue = fmt.Sprint(v)
+						}
+						tags[key] = stringValue
+						fields["_"+key] = value
```

Comment:
It would increase the size of the measurement a lot. Unlike tags which is de-duplicated compound by nature, fields are are stored line by line. If the key is designated as dimension, it's probably safe to just store it as tag.